Global CB Support

tt-aho opened the pull request:
### Ticket
https://github.com/tenstorrent/tt-metal/issues/14108

### Problem description
Metal infra support needed for creating and dispatching global circular buffers.

### What's changed
Add metal infra support for creating and dispatching global circular buffers.

- Split CB constants from circular_buffer.h to avoid building kernel file when building metal lib
- Add GlobalCircularBuffer implementation. a Global CB has 2 buffers
  - cb_buffer is the actual buffer space in L1 that is used for pushing/popping data
  - cb_config_buffer is the data needed to populate the remote cb interface when initializing/using remote cbs in kernels
  - config contains 2 arrays of ptrs (l1 aligned) used for tiles acked/sent sized based on num_receivers_per_sender
- Add a RemoteCBInterface struct, that is a union with the original CBInterface (renamed to LocalCBInterface) struct in order to save memory when creating the sized array. Each cb_id in the array corresponds to either a local or remote config, not both. We assert the two structs are the same size as we provide pointer aliases to the array to simplify access for whether it's a local/remote cb
- Add new apis/functions for initializing the remote interfaces, as well as apis to interact/use the remote cbs in a new remote_circular_buffer_api.h file taken from the original remote cb microbenchmarks
- Add support in dispatch for dispatching remote cb configs. Currently the config is 1 word representing the address to the config buffer to read/use
- Update cb dispatch to support dispatching remote cbs. Kernel launch msg now contains a remote cb offset, as well as the min_start_remote_cb_id. 
Dispatch data is laid out as follows:
Local CB data from 0 to max index across all cores, followed by remote CB data from 31 to min index across all cores.
We require that the min start remote index >= max end local index to avoid overlaps with initialization, since local cbs are sequentially initialized from 0 up, whereas remote cbs are sequentially initialized from 31 down. This also means for optimal dispatch performance, users should specify remote indices sequentially from 31, otherwise we are sending and initializing extraneous data, similarly to how local cb indices should go from 0 for optimal perf.

We decided to add the new global circular buffer support under an experimental namespace.
On the kernel side we are constrained with local memory size for the statically allocated interface array, so we cannot store as much data as we could for potentially simplifying user experience. Ex user must manually call secondary initialization functions in their own kernel, which is fairly error prone.
On the host side, we kept almost all apis the same except for a new CircularBufferConfig constructor to minimize api changes, however this may not be the best approach and potentially new apis should be added instead. TBD until we finalize some other cb api aspects regarding things such as potentially autogenerating ids.

### Checklist
- [x] Post commit CI passes https://github.com/tenstorrent/tt-metal/actions/runs/12142464308
- [ ] Blackhole Post commit (if applicable)
- [x] Model regression CI testing passes (if applicable) https://github.com/tenstorrent/tt-metal/actions/runs/12141520196
- [ ] Device performance regression CI testing passes (if applicable)
- [x] New/Existing tests provide coverage for changes

pgkeller commented on tt_metal/hw/firmware/src/ncrisc.cc:
```diff
@@ -95,8 +95,14 @@ int main(int argc, char *argv[]) {
         uint32_t tt_l1_ptr *cb_l1_base = (uint32_t tt_l1_ptr *)(kernel_config_base +
-            launch_msg->kernel_config.cb_offset);
-        setup_cb_read_write_interfaces(cb_l1_base, 0, launch_msg->kernel_config.max_cb_index, true, true, false);
+            launch_msg->kernel_config.local_cb_offset);
```
threshold doesn't mean much to me - threshold to what?  I think max may be better?
this sequence shows up a few places, wonder if it should move to firmware_common code

tt-aho replied:
I named it threshold since I reused the variable for the local and remote cb setup, so it's contains the max local cb index, then the min remote cb index. I can rename/separate them

pgkeller commented on tt_metal/hw/firmware/src/ncrisc.cc:
```diff
@@ -33,7 +33,7 @@ uint32_t noc_nonposted_writes_acked[NUM_NOCS] __attribute__((used));
 uint32_t noc_posted_writes_num_issued[NUM_NOCS] __attribute__((used));
 
-CBInterface cb_interface[NUM_CIRCULAR_BUFFERS] __attribute__((used));
+CBInterface local_remote_cb_interface[NUM_CIRCULAR_BUFFERS] __attribute__((used));
```
is this left over from the initial version w/ 3 pointers?  since there is no local/remote distinction, I think cb_interface is a better name

tt-aho replied:
Yes, will rename back to cb_interface

pgkeller commented on tt_metal/hw/inc/circular_buffer.h:
```diff
@@ -75,24 +93,51 @@ inline void setup_cb_read_write_interfaces(uint32_t tt_l1_ptr *cb_l1_base, uint3
+    for (uint32_t cb_id = NUM_CIRCULAR_BUFFERS - 1, end_id = start_cb_index - 1; cb_id != end_id; cb_id--) {
+        volatile tt_l1_ptr uint32_t* l1_remote_cb_config_addr = reinterpret_cast<volatile tt_l1_ptr uint32_t*>(circular_buffer_config_addr[0]);
+        RemoteCBInterface& cb_interface = get_remote_cb_interface(cb_id);
+        cb_interface.fifo_start_addr = l1_remote_cb_config_addr[0];
```
since l1_remote_cb_config_addr is a volatile, this sequence will look like:
load store
load store
load store
and therefore pay the L1 latency cost on each load.  Instead, do the loads up front into temporaries and then the stores to hide the L1 cost.  (and confirm before/after in the assembly)

yugaoTT reviewed (commented):
when running the subdevice ops, we probably need to update the compute kernel pointers as well (inplace CB update to the Global CB pointer). I wonder should this be done within the kernel or have a better way of hiding it.

tt-aho committed:
#14108: Add global CB implementation

tt-aho committed:
#14108: Update remote cb sync tests to use global cbs
Update GlobalSemaphore tests to use DispatchFixture
Add global circular buffer tests

tt-aho committed:
#14108: Add support for dispatching/initializing global CBs

eyonland commented on tt_metal/impl/buffers/circular_buffer_types.hpp:
```diff
@@ -50,29 +50,39 @@ class CircularBufferConfig {
+    const std::unordered_set<uint8_t> &local_buffer_indices() const;
+    const std::unordered_set<uint8_t> &remote_buffer_indices() const;
+
     const std::array<std::optional<tt::DataFormat>, NUM_CIRCULAR_BUFFERS> &data_formats() const;
```
Just a suggestion here...
Instead of adding these sets you could instead introduce 
```
struct ConfigData {
    tt::DataFormat data_format;
    uint32_t page_size;
    std::optional<Tile> tile;
    bool is_remote;
};
// With only 32 a std::array is nice.  
std::array<ConfigData, NUM_CIRCULAR_BUFFERS> configs_{};
// alternatively
std::array<std::optional<ConfigData>, NUM_CIRCULAR_BUFFERS> configs_{};
```

eyonland replied:
I think it might be better but this was just a suggestion though.  Up to you.

tt-aho committed:
#14108: Bump NCRISC FW size due to .text overflow on GS with profiler after integrating remote cb initialization in fw

rdjogoTT commented on tt_metal/hw/ckernels/blackhole/metal/llk_api/llk_pack_api.h:
```diff
@@ -213,19 +214,20 @@ inline void llk_pack_untilize(
     for (std::uint32_t block_rt = 0; block_rt < block_rt_dim; block_rt++) {
-        _llk_pack_untilize_<block_ct_dim, full_ct_dim, diagonal>(
-            pack_tile_addr, pack_dst_format[output_id], face_r_dim, num_faces, block_rt * block_ct_dim);
+        for (std::uint32_t block_rt = 0; block_rt < block_rt_dim; block_rt++) {
```
looks likes something weird happened here

tt-aho replied:
😞 thought I fixed this already. Thanks for catching

tt-aho replied:
Fixed

tt-aho commented on tt_metal/hw/inc/debug/sanitize_noc.h:
```diff
@@ -414,7 +414,7 @@ void debug_sanitize_noc_and_worker_addr(
             ((uint64_t)NOC_CMD_BUF_READ_REG(noc_id, NCRISC_WR_CMD_BUF, NOC_TARG_ADDR_MID) << 32) | noc_a_lower, \
         worker_a,                                                                                               \
-        NOC_CMD_BUF_READ_REG(noc_id, NCRISC_WR_CMD_BUF, NOC_AT_LEN_BE));
+        1);
```
This will be reverted/removed. Was changed for testing due to this issue https://github.com/tenstorrent/tt-metal/issues/15338

tt-aho replied:
Reverted

tt-aho committed:
#14108: Move global CBs to v1 namespace

tt-aho commented on tt_metal/host_api.hpp:
```diff
@@ -42,6 +42,12 @@ class Event;
+namespace experimental {
+
+class GlobalCircularBuffer;
+
```
I've moved it to v1 namespace, please take another look.

pavlejosipovic reviewed (approved):
From conv/max_pool perspective this looks good.
